File: SenLib/Sen3/FileFixes/book06_dat.cs
using HyoutaUtils;
using HyoutaUtils.Streams;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace SenLib.Sen3.FileFixes {
	class book06_dat : FileMod {
		public string GetDescription() {
			return "Fix typos in 'Other' books.";
		}

		public IEnumerable<FileModResult> TryApply(FileStorage storage) {
			var file = storage.TryGetDuplicate(new HyoutaUtils.Checksum.SHA1(0xfa7e69755a3592cdul, 0xb5196cd45ac86167ul, 0x1798384cu));
			if (file == null) {
				return null;
			}

			var bin = file.CopyToMemory();
			var book = new BookTable(bin, EndianUtils.Endianness.LittleEndian);

			// provincal -> provincial
			book.Entries[24].Text = book.Entries[24].Text.Insert(93, "i");

			var bin2 = book.WriteToStream(EndianUtils.Endianness.LittleEndian);
			return new FileModResult[] { new FileModResult("data/scripts/book/dat_en/book06.dat", bin2) };
		}

		public IEnumerable<FileModResult> TryRevert(FileStorage storage) {
			var file = storage.TryGetDuplicate(new HyoutaUtils.Checksum.SHA1(0xfa7e69755a3592cdul, 0xb5196cd45ac86167ul, 0x1798384cu));
			if (file == null) {
				return null;
			}
			return new FileModResult[] { new FileModResult("data/scripts/book/dat_en/book06.dat", file) };
		}
	}
}

File: native/sen3/FileFixes/book06_dat.cpp
#include <filesystem>
#include <string_view>
#include <vector>

#include "file_getter.h"

#include "p3a/pack.h"
#include "p3a/structs.h"
#include "sen/book_table.h"
#include "sha1.h"
#include "util/stream.h"

namespace SenLib::Sen3::FileFixes::book06_dat {
std::string_view GetDescription() {
    return "Fix typos in 'Other' books.";
}

bool TryApply(const SenPatcher::GetCheckedFileCallback& getCheckedFile,
              std::vector<SenPatcher::P3APackFile>& result) {
    try {
        auto file = getCheckedFile(
            "data/scripts/book/dat_en/book06.dat",
            18830,
            SenPatcher::SHA1FromHexString("fa7e69755a3592cdb5196cd45ac861671798384c"));
        if (!file) {
            return false;
        }

        DuplicatableByteArrayStream bin(file->Data.data(), file->Data.size());
        BookTable book(bin, HyoutaUtils::EndianUtils::Endianness::LittleEndian);

        // provincal -> provincial
        auto& entry24 = book.Entries.at(24).Text.value();
        entry24.insert(entry24.begin() + 93, 'i');

        std::vector<char> bin2;
        {
            MemoryStream ms(bin2);
            book.WriteToStream(ms, HyoutaUtils::EndianUtils::Endianness::LittleEndian);
        }

        result.emplace_back(std::move(bin2), file->Filename, SenPatcher::P3ACompressionType::LZ4);

        return true;
    } catch (...) {
        return false;
    }
}
} // namespace SenLib::Sen3::FileFixes::book06_dat
